src/udpreplay.cpp

```diff
@@ -182,21 +200,46 @@ int main(int argc, char *argv[]) {
                                                   ip->ip_hl * 4);
       if (interval != -1) {
         // Use constant packet rate
-        usleep(interval * 1000);
+        deadline.tv_sec += interval / 1000L;
+        deadline.tv_nsec += interval * 1000000L;
       } else {
-        if (tv.tv_sec == 0) {
-          tv = header.ts;
+        // Next packet deadline = start + (packet ts - first packet ts) * speed
+        int64_t delta =
+            (header.ts.tv_sec - pcap_start.tv_sec) * NANOSECONDS_PER_SECOND +
+            (header.ts.tv_usec -
+             pcap_start.tv_nsec); // Note PCAP_TSTAMP_PRECISION_NANO
+        if (speed != 1.0) {
+          delta *= speed;
+        }
+        deadline = start;
+        deadline.tv_sec += delta / NANOSECONDS_PER_SECOND;
+        deadline.tv_nsec += delta % NANOSECONDS_PER_SECOND;
+      }
+
+      // Normalize timespec
+      if (deadline.tv_nsec > NANOSECONDS_PER_SECOND) {
+        deadline.tv_sec++;
+        deadline.tv_nsec -= NANOSECONDS_PER_SECOND;
+      }
+
+      timespec now = {};
+      if (clock_gettime(CLOCK_MONOTONIC, &now) == -1) {
+        std::cerr << "clock_gettime: " << strerror(errno) << std::endl;
+        return 1;
+      }
```

Comment:
I am not sure if I really like/agree with this change. In my experience having more than 1 calls to the clock_gettime() in a precise loop is a good way to shoot itself in a foot.  You've gone from 1 to 3 for no good reason. Unless you are overloading your system, that check deadline > now is always false, so you are just wasted some CPU cycles for calling clock_gettime() and doing useless computations on every packet. clock_nanosleep() below would just be a NOP() if we are running behind.

Reply:
No, there is only 1 call except the first loop which is 3. `clock_gettime` takes about ~40ns since it's a vDSO / vsyscall on any reasonable kernel, but `clock_nanosleep` is always a syscall with ~5us overhead in my system. So eliding `clock_nanosleep` if possible improves accuracy. Further more `clock_nanosleep` seems to have a jitter of 60us on my system if the deadline is in the future. I think that is due to the scheduler being invoked and granularity of the timer wheels in the kernel. So your code will smear out bursts of packets while my version would introduce a overhead less than 100ns plus the overhead of `sendto`.

Test code:
```cpp
#include <chrono>
#include <cstring>
#include <iostream>

int main(int argc, char *argv[]) {

  int iter = 1000000;
  timespec deadline = {};

  auto start = std::chrono::steady_clock::now();

  for (int i = 0; i < iter; ++i) {
    // if (clock_gettime(CLOCK_MONOTONIC, &deadline) == -1) {
    //   std::cerr << "clock_gettime: " << strerror(errno) << std::endl;
    //   return 1;
    // }
    if (clock_nanosleep(CLOCK_MONOTONIC, TIMER_ABSTIME, &deadline, nullptr) ==
        -1) {
      std::cerr << "clock_nanosleep: " << strerror(errno) << std::endl;
      return 1;
    }
  }

  auto end = std::chrono::steady_clock::now();
  auto diff = (end - start).count() / iter;

  std::cout << diff << std::endl;

  return 0;
}
```